src/relay/backend/te_compiler.cc

```diff
@@ -349,6 +350,9 @@ class LowerTensorExpr : public ExprMutator {
       Map<GlobalVar, tir::PrimFunc> prim_fns;
 
       for (auto prim_fn : ext_func->funcs->functions) {
+        if (prim_fn.second->GetAttr<String>(attr::kCompiler).defined()) {
```

Comment:
Just querying, when is this untrue for an external function that's gone via `Lower` ?

Reply:
Hi @Mousius, After calling `TECompilerImpl::Lower()` on an external function, the TECompiler will just encapsulate it into `CachedFunc` and return, and the external function is not lowered. It will be lowered by the external codegen with the `TECompilerNode::LowerExternalFunctions()` later. I added a line `ir_module->Add(global_var, key->source_func);` to `LowerInternal`, so this is untrue when some function in the `funcs` of the `CachedFuncNode` is not lowered yet.

Reply:
Hi @YuchenJin, thanks for getting back to me, the reason I asked was because this loop is gated by the `kCompiler` attribute:
```
if (func->GetAttr<String>(attr::kCompiler).defined()) {
```
And then when it goes into `LowerInternal` it'll end up in this block (with your alterations):
```
if (key->source_func->GetAttr<String>(attr::kCompiler).defined()) {
      auto ir_module = IRModule();
      const auto name_node = key->source_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
      ICHECK(name_node.defined()) << "External function has not been attached a name yet.";
      auto func_name = GetUniqueName(name_node.value(), &name_map_);
      auto target = Target("ext_dev");
      auto global_var = GlobalVar(func_name);
      global_var->checked_type_ = key->source_func->checked_type();
      ir_module->Add(global_var, key->source_func);
      value->cached_func = CachedFunc(target, global_var, {}, {}, te::Schedule(), {}, ir_module);
      return value;
    }
```
Which means when we enter this loop, the only function in the `IRModule` should be a function taken from `key->source_func` which has the attr on it?

Reply:
Thanks @Mousius for elaborating it! I think you are right, and we can probably remove this loop because the `IRModule` should not contain a PrimFunc anyway for external functions, and we don't need to check it. What do you think?

Reply:
Let's try removing the loop, if tests still pass we should be confident it wasn't necessary :smile_cat:

Reply:
Yeah, tests still pass after removing it. @jroesch, could you review this PR? :)